bin/npg_add_illumina_composition_metadata.pl

```diff
@@ -0,0 +1,431 @@
+    next;
+  }
+
+  my $composition_json = $meta{$COMPOSITION};
+  my $composition;
+  if ($composition_json) {
+    $composition = npg_tracking::glossary::composition->thaw(
+      $composition_json, component_class =>
+      'npg_tracking::glossary::composition::component::illumina');
+  } else {
+    my $method = $path =~ m{/illumina/runs}smx ? 'logcroak' : 'debug';
+    $logger->$method('No composition metadata for ', $path);
+
+    my %attrs = ();
+    for my $attr ($ID_RUN, $POSITION, $TAG_INDEX, $ALIGNMENT_FILTER) {
+      if (defined $meta{$attr}) {
+        if ($attr eq $ALIGNMENT_FILTER) {
+          $attrs{'subset'} = $meta{$attr};
+        } elsif ($attr eq $POSITION) {
+          $attrs{'position'} = $meta{$attr};
+	} else {
+          $attrs{$attr} = $meta{$attr};
+        }
+      }
+    }
+
+    my ($ifile_name, $idir) = fileparse $path;
+
+    my $file_name_mismatch = 0;
+    my $subset = q[];
+    my $subset_mismatch = 0;
+
+    ($attrs{'id_run'} && $attrs{'position'}) or $logger->logcroak(
+      "Either id_run or position is not defined for $path");
+
+    $composition = npg_tracking::glossary::composition->new(
+      components => [
+        npg_tracking::glossary::composition::component::illumina->new(\%attrs)
+      ]
+    );
+    my $product = npg_pipeline::product->new(composition => $composition);
+    my $expected_name = $product->file_name_root(); #? with subset
+    if ($ifile_name !~ /\A$expected_name\.(?bam|cram)\Z/xms) {
+      $logger->warn('File name mismatch: ', $path, ' and ', $expected_name);
+      $file_name_mismatch = 1;
+    }
+
+    if ($attrs{'subset'}) {
+      $subset = $attrs{'subset'};
+      if ($ifile_name !~ /$subset/smx) {
+        $logger->error(
+          "Subset $subset mismatch with file name in $path, skipping");
+        $subset_mismatch = 1;
+      }
+    }
+
+    if ($report_path and ($file_name_mismatch or $subset_mismatch)) {
+      my $line = join q[,], $path, $file_name_mismatch,
+                            $subset, $subset_mismatch;
+      print $report_fh $line or $logger->logcroak(
+          'Failed to write to ' , $report_path, q[: ], $OS_ERROR);
+    }
+
+    $composition_json = $composition->freeze();
+    $logger->info("Generated composition for ${path}: $composition_json");
+
+    if ($subset_mismatch) {
+      $num_mismatches++;
+      next;
+    }
+
+    # Create a composition file.
+    my $cpath = $idir . $expected_name . q[.composition.json];
+    try {
+      $num_created++;
+      $dry_run or
+      _create_irods_composition_file($cpath, $composition_json);
+    } catch {
+      $logger->logcroak("Failed to create file ${cpath}: $_");
+    };
+    # Will not deal with ACLs for the new file here, the metadata updater
+    # can fix them later if necessary. The same for secondary metadata.
+    # Setting the same metadata on a file gives an error.
+    try {
+      $irods->add_object_avu($cpath, 'type', 'json');
```

Comment:
If you use `WTSI::NPG::iRODS::DataObject::add_avu` you can avoid the try-catch, if you like, because it checks for you.